pkg/clients/eks/eks.go

```diff
@@ -19,702 +19,335 @@ package eks
+		}
+		if len(in.ResourcesVpcConfig.SubnetIDs) == 0 && len(cluster.ResourcesVpcConfig.SubnetIds) > 0 {
+			in.ResourcesVpcConfig.SubnetIDs = cluster.ResourcesVpcConfig.SubnetIds
+		}
 	}
+	in.RoleArn = awsclients.LateInitializeStringPtr(in.RoleArn, cluster.RoleArn)
+	in.Version = awsclients.LateInitializeStringPtr(in.Version, cluster.Version)
+}
 
-	result := make([]*ec2.Image, len(out.Images))
-	for i := 0; i < len(out.Images); i++ {
-		result[i] = &out.Images[i]
+// IsUpToDate checks whether there is a change in any of the modifiable fields.
+func IsUpToDate(p *v1beta1.ClusterParameters, cluster *eks.Cluster) (bool, error) {
+	patch, err := CreatePatch(cluster, p)
+	if err != nil {
+		return false, err
 	}
 
-	return result, nil
-}
-
-func getMostRecentImage(images []*ec2.Image) *ec2.Image {
-	var result *ec2.Image
-	for i := 0; i < len(images); i++ {
-		if result == nil || aws.StringValue(result.CreationDate) < aws.StringValue(images[i].CreationDate) {
-			result = images[i]
+	// NOTE(hasheddan): AWS removes insignificant bits from CIDRs, so we must
+	// compare by converting user-supplied CIDRs to network blocks. We only skip
+	// comparison if both external and local have no CIDR blocks defined.
+	if (cluster.ResourcesVpcConfig != nil && len(cluster.ResourcesVpcConfig.PublicAccessCidrs) > 0) || len(p.ResourcesVpcConfig.PublicAccessCidrs) > 0 {
+		// Convert user-supplied slice of CIDRs to map of networks.
+		netMap := map[string]bool{}
+		for _, c := range p.ResourcesVpcConfig.PublicAccessCidrs {
+			_, net, err := net.ParseCIDR(c)
+			if err != nil {
+				return false, err
+			}
+			netMap[net.String()] = true
+		}
+		// If length of networks does not match the length of CIDR blocks
+		// returned by AWS then we need update.
+		if len(netMap) != len(cluster.ResourcesVpcConfig.PublicAccessCidrs) {
+			return false, nil
+		}
+		// If AWS returns a CIDR block that is not in the map, then we need
+		// update.
+		for _, pc := range cluster.ResourcesVpcConfig.PublicAccessCidrs {
+			if !netMap[pc] {
+				return false, nil
+			}
 		}
 	}
 
-	return result
+	return cmp.Equal(&v1beta1.ClusterParameters{}, patch, cmpopts.EquateEmpty(),
+		cmpopts.IgnoreTypes(&v1alpha1.Reference{}, &v1alpha1.Selector{}),
+		cmpopts.IgnoreFields(v1beta1.ClusterParameters{}, "Tags"),
+		cmpopts.IgnoreFields(v1beta1.VpcConfigRequest{}, "SecurityGroupIDRefs", "SubnetIDRefs", "PublicAccessCidrs")), nil
 }
 
-func getImageWithID(imgName string, images []*ec2.Image) (*ec2.Image, error) {
-	for _, img := range images {
-		if imgName == aws.StringValue(img.ImageId) {
-			return img, nil
-		}
+// GetConnectionDetails extracts managed.ConnectionDetails out of eks.Cluster.
+func GetConnectionDetails(cluster *eks.Cluster, stsClient STSClient) managed.ConnectionDetails {
+	if cluster == nil || cluster.Name == nil || cluster.Endpoint == nil || cluster.CertificateAuthority == nil || cluster.CertificateAuthority.Data == nil {
```

Comment:
This is a pretty verbose check here, and while working on this controller I notice that we make assumptions in a lot of places that we will not get `nil` values for subfields when the parent is not `nil`. In practice this works fine because AWS returns these values or we check an error before dereferencing all the time. However, it makes me a little uncomfortable that we don't always check every `nil` pointer. In this case, I do this check because I don't want to write a connection details until I have everything we need for a valid `kubeconfig`.

Reply:
We should always check for `nil` if the type is a pointer type. Are there places in other parts of the codebase that you realized we don't check for `nil` even though we should?

Reply:
Yes, and I believe we had a discussion about this on a previous issue. This isn't exactly checking for `nil`, but we have a few areas like this where we might, for example, not explicitly check that we don't have an array out of bounds  https://github.com/crossplane/provider-aws/blob/3238aeef3ccc0dab5672482577749432d2d60082/pkg/controller/database/rdsinstance.go#L136

Reply:
Yeah, this looks like something to update some time. It's not critical probably because it's meaninglessly an array; if `NotFound` it returns error anyway and it can never be `>1` elements since we give exact name but yeah, why not check if there is a possibility.